File: screenstudio-agent/tool/src/session.ts
import { spawn } from "node:child_process"
import { constants } from "node:fs"
import { promises as fs } from "node:fs"
import * as path from "node:path"
import { EventsDoc, RunSummary, StudioSession, TimelineEvent } from "./types.js"
import { ensureTool, exists, nowIso, PROJECT_ROOT, readJson, StudioError, timestampName, writeJson } from "./util.js"

const STATE_DIR_NAME = ".agentic-studio"
const CURRENT_FILE = "current.json"

export interface SessionPaths {
  root: string
  stateDir: string
  currentFile: string
  runDir: string
  rawVideo: string
  eventsFile: string
  sessionFile: string
  recorderLog: string
}

export interface StartOptions {
  name?: string
  outDir?: string
  display?: number | null
  audio?: boolean
  cursor?: boolean
  showClicks?: boolean
  duration?: number | null
}

export function projectPaths(root: string, runDir?: string): SessionPaths {
  const stateDir = path.join(root, STATE_DIR_NAME)
  const actualRunDir = runDir ?? path.join(root, "runs", "manual")
  return {
    root,
    stateDir,
    currentFile: path.join(stateDir, CURRENT_FILE),
    runDir: actualRunDir,
    rawVideo: path.join(actualRunDir, "raw.mov"),
    eventsFile: path.join(actualRunDir, "events.json"),
    sessionFile: path.join(actualRunDir, "session.json"),
    recorderLog: path.join(actualRunDir, "recorder.log"),
  }
}

export function buildScreencaptureCommand(rawVideo: string, opts: StartOptions): string[] {
  const args = ["-v", "-x"]
  if (opts.cursor ?? true) args.push("-C")
  if (opts.showClicks ?? true) args.push("-k")
  if (opts.audio) args.push("-g")
  if (opts.display !== undefined && opts.display !== null) args.push(`-D${opts.display}`)
  if (opts.duration !== undefined && opts.duration !== null) args.push(`-V${opts.duration}`)
  args.push(rawVideo)
  return args
}

export function isProcessRunning(pid: number): boolean {
  try {
    process.kill(pid, 0)
    return true
  } catch {
    return false
  }
}

export async function loadActiveSession(root = PROJECT_ROOT, required = true): Promise<StudioSession | null> {
  const currentFile = projectPaths(root).currentFile
  if (!(await exists(currentFile))) {
    if (required) throw new StudioError("No active recording session found.")
    return null
  }
  const pointer = await readJson<{ session_file: string }>(currentFile)
  if (!(await exists(pointer.session_file))) {
    if (required) throw new StudioError(`Active session metadata is missing: ${pointer.session_file}`)
    return null
  }
  return readJson<StudioSession>(pointer.session_file)
}

export async function resolveSession(root = PROJECT_ROOT, runDir?: string): Promise<StudioSession> {
  if (!runDir) {
    const active = await loadActiveSession(root, true)
    if (!active) throw new StudioError("No active recording session found.")
    return active
  }
  const sessionFile = path.join(runDir, "session.json")
  if (!(await exists(sessionFile))) throw new StudioError(`No session.json found in ${runDir}`)
  return readJson<StudioSession>(sessionFile)
}

export async function startSession(root = PROJECT_ROOT, opts: StartOptions = {}): Promise<StudioSession> {
  await ensureTool("screencapture")
  const active = await loadActiveSession(root, false)
  if (active && isProcessRunning(active.pid)) {
    throw new StudioError(`A recording is already active in ${active.run_dir} with pid ${active.pid}.`)
  }

  const runName = opts.name || timestampName()
  const baseOut = path.resolve(root, opts.outDir || "runs")
  const runDir = path.join(baseOut, runName)
  const paths = projectPaths(root, runDir)
  await fs.mkdir(paths.runDir, { recursive: true })
  await fs.mkdir(paths.stateDir, { recursive: true })
  if (await exists(paths.rawVideo)) throw new StudioError(`Refusing to overwrite existing video: ${paths.rawVideo}`)

  const logHandle = await fs.open(paths.recorderLog, "a")
  const startedAt = nowIso()
  const startEpoch = Date.now() / 1000
  const child = spawn("screencapture", buildScreencaptureCommand(paths.rawVideo, opts), {
    cwd: root,
    detached: true,
    stdio: ["ignore", logHandle.fd, logHandle.fd],
  })
  child.unref()
  await new Promise((resolve) => setTimeout(resolve, 700))
  await logHandle.close()
  if (child.exitCode !== null) {
    throw new StudioError(`screencapture exited immediately with code ${child.exitCode}. See ${paths.recorderLog}`)
  }

  const session: StudioSession = {
    version: 1,
    pid: child.pid ?? 0,
    root,
    run_dir: paths.runDir,
    raw_video: paths.rawVideo,
    events_file: paths.eventsFile,
    session_file: paths.sessionFile,
    recorder_log: paths.recorderLog,
    started_at: startedAt,
    start_epoch: startEpoch,
    record: {
      display: opts.display ?? null,
      audio: !!opts.audio,
      cursor: opts.cursor ?? true,
      show_clicks: opts.showClicks ?? true,
      duration: opts.duration ?? null,
    },
    status: "recording",
  }
  const events: EventsDoc = {
    version: 1,
    recording: { started_at: startedAt, start_epoch: startEpoch, raw_video: paths.rawVideo },
    events: [],
  }
  await writeJson(paths.sessionFile, session)
  await writeJson(paths.eventsFile, events)
  await writeJson(paths.currentFile, { session_file: paths.sessionFile })
  return session
}

export async function appendEvent(root = PROJECT_ROOT, event: TimelineEvent, runDir?: string): Promise<TimelineEvent> {
  const session = await resolveSession(root, runDir)
  const doc = await readJson<EventsDoc>(session.events_file)
  const normalized = { ...event }
  if (normalized.time === undefined) {
    normalized.time = Math.max(0, Date.now() / 1000 - doc.recording.start_epoch)
    normalized.time = Math.round(normalized.time * 1000) / 1000
  }
  normalized.created_at ??= nowIso()
  doc.events ??= []
  doc.events.push(normalized)
  doc.events.sort((a, b) => Number(a.time || 0) - Number(b.time || 0))
  await writeJson(session.events_file, doc)
  return normalized
}

export async function stopSession(root = PROJECT_ROOT, timeout = 20): Promise<StudioSession> {
  const session = await resolveSession(root)
  if (session.status !== "stopped" && isProcessRunning(session.pid)) {
    try {
      process.kill(-session.pid, "SIGINT")
    } catch {
      // ignore
    }
    const deadline = Date.now() + timeout * 1000
    while (Date.now() < deadline && isProcessRunning(session.pid)) {
      await new Promise((resolve) => setTimeout(resolve, 200))
    }
    if (isProcessRunning(session.pid)) {
      try {
        process.kill(-session.pid, "SIGTERM")
      } catch {
        // ignore
      }
    }
  }

  const stoppedAt = nowIso()
  const stopEpoch = Date.now() / 1000
  session.status = "stopped"
  session.stopped_at = stoppedAt
  session.stop_epoch = stopEpoch
  await writeJson(session.session_file, session)

  if (await exists(session.events_file)) {
    const doc = await readJson<EventsDoc>(session.events_file)
    doc.recording.stopped_at = stoppedAt
    doc.recording.stop_epoch = stopEpoch
    doc.recording.duration = Math.round((stopEpoch - doc.recording.start_epoch) * 1000) / 1000
    await writeJson(session.events_file, doc)
  }

  const currentFile = projectPaths(root).currentFile
  if (await exists(currentFile)) {
    const pointer = await readJson<{ session_file?: string }>(currentFile)
    if (pointer.session_file === session.session_file) await fs.unlink(currentFile)
  }
  return session
}

export async function listRuns(root = PROJECT_ROOT): Promise<RunSummary[]> {
  const runsDir = path.join(root, "runs")
  if (!(await exists(runsDir))) return []
  const entries = await fs.readdir(runsDir, { withFileTypes: true })
  const dirs = entries.filter((entry) => entry.isDirectory()).map((entry) => path.join(runsDir, entry.name))
  const withTimes = await Promise.all(dirs.map(async (dir) => ({ dir, stat: await fs.stat(dir) })))
  withTimes.sort((a, b) => b.stat.mtimeMs - a.stat.mtimeMs)
  const runs: RunSummary[] = []
  for (const { dir } of withTimes) {
    const eventsFile = path.join(dir, "events.json")
    let events = 0
    if (await exists(eventsFile)) {
      try {
        events = (await readJson<EventsDoc>(eventsFile)).events?.length ?? 0
      } catch {
        events = 0
      }
    }
    runs.push({
      name: path.basename(dir),
      run_dir: dir,
      raw: await exists(path.join(dir, "raw.mov")),
      final: await exists(path.join(dir, "final.mp4")),
      events,
      session: await exists(path.join(dir, "session.json")),
    })
  }
  return runs
}

File: screenstudio-agent/studio_agent/session.py
from __future__ import annotations

import os
import signal
import subprocess
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from .util import StudioError, ensure_tool, now_iso, read_json, timestamp_name, write_json


STATE_DIR_NAME = ".agentic-studio"
CURRENT_FILE = "current.json"


@dataclass(frozen=True)
class SessionPaths:
    root: Path
    state_dir: Path
    current_file: Path
    run_dir: Path
    raw_video: Path
    events_file: Path
    session_file: Path
    recorder_log: Path


def project_paths(root: Path, run_dir: Path | None = None) -> SessionPaths:
    state_dir = root / STATE_DIR_NAME
    actual_run_dir = run_dir or root / "runs" / "manual"
    return SessionPaths(
        root=root,
        state_dir=state_dir,
        current_file=state_dir / CURRENT_FILE,
        run_dir=actual_run_dir,
        raw_video=actual_run_dir / "raw.mov",
        events_file=actual_run_dir / "events.json",
        session_file=actual_run_dir / "session.json",
        recorder_log=actual_run_dir / "recorder.log",
    )


def build_screencapture_command(
    raw_video: Path,
    *,
    display: int | None = None,
    audio: bool = False,
    cursor: bool = True,
    show_clicks: bool = True,
    duration: float | None = None,
) -> list[str]:
    cmd = ["screencapture", "-v", "-x"]
    if cursor:
        cmd.append("-C")
    if show_clicks:
        cmd.append("-k")
    if audio:
        cmd.append("-g")
    if display is not None:
        cmd.append(f"-D{display}")
    if duration is not None:
        cmd.append(f"-V{duration:g}")
    cmd.append(str(raw_video))
    return cmd


def is_process_running(pid: int) -> bool:
    try:
        os.kill(pid, 0)
    except OSError:
        return False
    return True


def load_active_session(root: Path, *, required: bool = True) -> dict[str, Any] | None:
    current = project_paths(root).current_file
    if not current.exists():
        if required:
            raise StudioError("No active recording session found.")
        return None
    pointer = read_json(current)
    session_file = Path(pointer["session_file"])
    if not session_file.exists():
        if required:
            raise StudioError(f"Active session metadata is missing: {session_file}")
        return None
    return read_json(session_file)


def resolve_session(root: Path, run_dir: Path | None = None) -> dict[str, Any]:
    if run_dir is None:
        session = load_active_session(root, required=True)
        assert session is not None
        return session
    session_file = run_dir / "session.json"
    if not session_file.exists():
        raise StudioError(f"No session.json found in {run_dir}")
    return read_json(session_file)


def start_session(
    root: Path,
    *,
    name: str | None = None,
    out_dir: Path | None = None,
    display: int | None = None,
    audio: bool = False,
    cursor: bool = True,
    show_clicks: bool = True,
    duration: float | None = None,
) -> dict[str, Any]:
    ensure_tool("screencapture")

    active = load_active_session(root, required=False)
    if active and is_process_running(int(active["pid"])):
        raise StudioError(
            f"A recording is already active in {active['run_dir']} with pid {active['pid']}."
        )

    run_name = name or timestamp_name()
    base_out = (out_dir or root / "runs").expanduser()
    run_dir = (base_out / run_name).resolve()
    paths = project_paths(root, run_dir)
    paths.run_dir.mkdir(parents=True, exist_ok=True)
    paths.state_dir.mkdir(parents=True, exist_ok=True)

    if paths.raw_video.exists():
        raise StudioError(f"Refusing to overwrite existing video: {paths.raw_video}")

    cmd = build_screencapture_command(
        paths.raw_video,
        display=display,
        audio=audio,
        cursor=cursor,
        show_clicks=show_clicks,
        duration=duration,
    )

    started_at = now_iso()
    start_epoch = time.time()
    log = paths.recorder_log.open("ab")
    proc = subprocess.Popen(
        cmd,
        stdin=subprocess.DEVNULL,
        stdout=log,
        stderr=log,
        start_new_session=True,
        close_fds=True,
    )
    time.sleep(0.7)
    if proc.poll() is not None:
        raise StudioError(
            f"screencapture exited immediately with code {proc.returncode}. "
            f"See {paths.recorder_log}"
        )

    session = {
        "version": 1,
        "pid": proc.pid,
        "root": str(root),
        "run_dir": str(paths.run_dir),
        "raw_video": str(paths.raw_video),
        "events_file": str(paths.events_file),
        "session_file": str(paths.session_file),
        "recorder_log": str(paths.recorder_log),
        "started_at": started_at,
        "start_epoch": start_epoch,
        "record": {
            "display": display,
            "audio": audio,
            "cursor": cursor,
            "show_clicks": show_clicks,
            "duration": duration,
        },
        "status": "recording",
    }
    events = {
        "version": 1,
        "recording": {
            "started_at": started_at,
            "start_epoch": start_epoch,
            "raw_video": str(paths.raw_video),
        },
        "events": [],
    }
    write_json(paths.session_file, session)
    write_json(paths.events_file, events)
    write_json(paths.current_file, {"session_file": str(paths.session_file)})
    return session


def append_event(root: Path, event: dict[str, Any], *, run_dir: Path | None = None) -> dict[str, Any]:
    session = resolve_session(root, run_dir)
    events_file = Path(session["events_file"])
    data = read_json(events_file)
    start_epoch = float(data["recording"]["start_epoch"])

    normalized = dict(event)
    if "time" not in normalized:
        normalized["time"] = round(max(0.0, time.time() - start_epoch), 3)
    normalized.setdefault("created_at", now_iso())

    data.setdefault("events", []).append(normalized)
    data["events"].sort(key=lambda item: float(item.get("time", 0)))
    write_json(events_file, data)
    return normalized


def stop_session(root: Path, *, run_dir: Path | None = None, timeout: float = 20.0) -> dict[str, Any]:
    session = resolve_session(root, run_dir)
    pid = int(session["pid"])

    if session.get("status") != "stopped" and is_process_running(pid):
        try:
            os.killpg(pid, signal.SIGINT)
        except ProcessLookupError:
            pass
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline and is_process_running(pid):
            time.sleep(0.2)
        if is_process_running(pid):
            os.killpg(pid, signal.SIGTERM)
            time.sleep(0.5)

    stopped_at = now_iso()
    stop_epoch = time.time()
    session["status"] = "stopped"
    session["stopped_at"] = stopped_at
    session["stop_epoch"] = stop_epoch
    write_json(Path(session["session_file"]), session)

    events_file = Path(session["events_file"])
    if events_file.exists():
        events = read_json(events_file)
        events.setdefault("recording", {})["stopped_at"] = stopped_at
        events.setdefault("recording", {})["stop_epoch"] = stop_epoch
        events.setdefault("recording", {})["duration"] = round(
            stop_epoch - float(events["recording"]["start_epoch"]), 3
        )
        write_json(events_file, events)

    current = project_paths(root).current_file
    if current.exists():
        pointer = read_json(current)
        if Path(pointer.get("session_file", "")) == Path(session["session_file"]):
            current.unlink()

    return session


def list_runs(root: Path) -> list[dict[str, Any]]:
    runs_dir = root / "runs"
    if not runs_dir.exists():
        return []
    runs: list[dict[str, Any]] = []
    for run_dir in sorted(runs_dir.iterdir(), key=lambda path: path.stat().st_mtime, reverse=True):
        if not run_dir.is_dir():
            continue
        events_file = run_dir / "events.json"
        session_file = run_dir / "session.json"
        final_file = run_dir / "final.mp4"
        raw_file = run_dir / "raw.mov"
        events_count = 0
        if events_file.exists():
            try:
                events_count = len(read_json(events_file).get("events", []))
            except Exception:
                events_count = 0
        runs.append(
            {
                "name": run_dir.name,
                "run_dir": str(run_dir),
                "raw": raw_file.exists(),
                "final": final_file.exists(),
                "events": events_count,
                "session": session_file.exists(),
            }
        )
    return runs
